canal/canal.go

```diff
@@ -154,14 +160,75 @@ func (c *Canal) run() error {
 
 	close(c.dumpDoneCh)
 
-	if err := c.startSyncBinlog(); err != nil {
+	syncBlock := func() error {
+		errSync := c.startSyncBinlog();
+
 		if !c.isClosed() {
-			log.Errorf("canal start sync binlog err: %v", err)
+			log.Errorf("canal start sync binlog err: %v", errSync)
 		}
-		return errors.Trace(err)
+		return errSync
+	}
+
+	return c.handleRetriableSync(syncBlock, c.prepareSyncer, syncErrorCh)
+}
+
+func (c *Canal) handleRetriableSync(syncF func() error, prepareF func() error, syncErrorCh chan error) error {
+	var attemptNum int
+	var sleep time.Duration
+
+	resetAttempts := func(a *int, s *time.Duration) {
+		*a = 0
+		*s = waitForRetrySeconds
+		c.proofOfSuccess = false
+	}
+
+	shouldStopRetrying := func(attemtp int, e error) bool {
+		if attemtp >= maxAttempts {
+			if e != nil {
+				// communicate the error to the upper layer (client)
+				// in case we exhaust all the attempts
+				resetAttempts(&attemptNum, &sleep)
+				syncErrorCh <- e
+				return true
+			}
+		}
+		return false
+	}
+
+	resetAttempts(&attemptNum, &sleep)
+
+	for {
+		errSync := syncF();
+
+		if c.proofOfSuccess {
```

Comment:
We can get rid of the 'proofOfSuccess' var if we always call 'resetAttempts' after each successful attempt. By automatically resetting the number of attempts and waitForRetrySeconds after each successful call, it will simplify the business logic and reduce the number of 'if' statements we'll need to account for.

Reply:
This is a test